Use JSON data file for storing PowerRename settings instead of registry

vldmr11080 opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
Use persistent json data file for storing PowerRename settings instead of registry. Currently whenever we call settings object to get / set certain setting we read / write from registry. Getting settings with new implementation won't require any IO operation, data is loaded from json formatted file on the startup. Settings are saved to file when certain settings change and when disabling PowerRenam / closing PowerToys. Migration of settings from registry is supported.

Most recently used list of search and replace strings will be implemented using json storage in follow up pull request.

**Example json file:**
```
{
    "ShowIcon": true,
    "ExtendedContextMenuOnly": false,
    "PersistState": true,
    "MRUEnabled": true,
    "MaxMRUSize": 10,
    "Flags": 0,
    "SearchText": "",
    "ReplaceText": ""
}
```

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #1206

vldmr11080 committed:
Use JSON data file for storing PowerRename settings instead of registry

yuyoyuppe commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -1,171 +1,60 @@
-    ULONG cb = cchBuf * sizeof(*value);
-    return (SUCCEEDED(HRESULT_FROM_WIN32(SHGetValue(HKEY_CURRENT_USER, c_rootRegPath, valueName, &type, value, &cb) == ERROR_SUCCESS)));
+    std::wstring GetRegString(const std::wstring& valueName) {
+        wchar_t* value = new wchar_t[CSettings::MAX_INPUT_STRING_LEN];
```
memory leak

yuyoyuppe commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,205 @@ void CRenameMRU::_FreeMRUList()
+    settings.replaceText             = GetRegString(c_replaceText);
+}
+
+void CSettings::ParseJsonSettings(const json::JsonObject& jsonSettings)
```
please remove this method, since you already initialize `Settings` class with the same values.

yuyoyuppe commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,205 @@ void CRenameMRU::_FreeMRUList()
+    std::scoped_lock lock{ dataLock };
+
+    auto jsonData = json::from_file(jsonFilePath);
+    return jsonData ? jsonData.value() : json::JsonObject();
```
there's no point in types if we convert everything in a string. please return `std::optional` and don't call `ParseJsonSettings`, since you already provide sane default values.

yuyoyuppe reviewed (changes requested):
This is only a brief review, will do a more detailed-one on Monday.
- why do we need mutex for `CSettings`? From the looks of it, we don't need it.
- `WindowsApp.lib` should be link with `PowerRenameLib`
- let's not link with `$(OutDir)` manually and use project references directly

yuyoyuppe commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,205 @@ void CRenameMRU::_FreeMRUList()
+{
+    std::scoped_lock lock{ dataLock };
+
+    settings.enabled                 = jsonSettings.GetNamedBoolean(c_enabled, true);
```
you already initialize `Settings` class with the same values, so you shouldn't provide any default values, just check their existence with `json::has`.

stefansjfw commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -1,171 +1,60 @@
-{
-    return (SUCCEEDED(HRESULT_FROM_WIN32(SHSetValue(HKEY_CURRENT_USER, c_rootRegPath, valueName, REG_DWORD, &value, sizeof(value)))));
-}
+    const wchar_t c_enabled[] = L"Enabled";
```
Should we be consistent with FZ settings and use dash-case instead of UpperCamelCase?

vldmr11080 replied:
It complicates migration a bit, since we need to handle different format, but it can be done.

enricogior replied:
Please keep the current UpperCamelCase, dash-case is just a bad idea to begin with and we will very likely remove it from FZ.

stefansjfw commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,205 @@ void CRenameMRU::_FreeMRUList()
+    {
+        ParseJsonSettings(GetPersistPowerRenameJSON());
+
+        SavePowerRenameData();
```
move SavePowerRenameData() out of if-else, so we save after migration to?

vldmr11080 replied:
It actually belongs to if part, not here in else (when we do migration than we save to json).

stefansjfw commented on src/modules/powerrename/lib/Settings.h:
```diff
@@ -1,45 +1,73 @@
+        bool MRUEnabled{ true };
+        long maxMRUSize{ 10 };
+        long flags{ 0 };
+        std::wstring searchText;
```
zero-initialize these two as well

stefansjfw replied:
as a follow up of default values comment above

vldmr11080 replied:
Done.

vldmr11080 committed:
Address PR comments and made several improvements

vldmr11080 committed:
Remove WindowsApp.lib dependencies in test app and unit tests

vldmr11080 committed:
Revert changes in vcxproj for unit test

vldmr11080 committed:
Solve linker warnings generated while linking WindowsApp.lib

vldmr11080 commented:
> This is only a brief review, will do a more detailed-one on Monday.
> 
> * why do we need mutex for `CSettings`? From the looks of it, we don't need it.
> * `WindowsApp.lib` should be link with `PowerRenameLib`
> * let's not link with `$(OutDir)` manually and use project references directly

Removed mutex and linked `WindowsApp.lib` with `PowerRenameLib`. Removed links to unit tests and test app.

enricogior commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,113 @@ void CRenameMRU::_FreeMRUList()
+    jsonData.SetNamedValue(c_searchText,              json::value(settings.searchText));
+    jsonData.SetNamedValue(c_replaceText,             json::value(settings.replaceText));
+
+    json::to_file(jsonFilePath, jsonData);
```
`json::to_file` can throw exceptions, right?

vldmr11080 replied:
No, `json::from_file` can, and it has try-catch block within it.

enricogior replied:
`ofstream` doesn't throw an exception if you pass an invalid file name or if it can't access the file?

vldmr11080 replied:
No, the do set error code, but no exception:
```
Constructs an ofstream object, initially associated with the file identified by its first argument (filename), open with the mode specified by mode.
Internally, its ostream base constructor is passed a pointer to a newly constructed filebuf object (the internal file stream buffer). Then, filebuf::open is called with filename and mode as arguments.
If the file cannot be opened, the stream's failbit flag is set.
```

enricogior commented on src/modules/powerrename/dll/dllmain.cpp:
```diff
@@ -282,13 +283,15 @@ class PowerRenameModule : public PowertoyModuleIface
     {
-        m_enabled = CSettings::GetEnabled();
+        CSettingsInstance().LoadPowerRenameData();
+        m_enabled = CSettingsInstance().GetEnabled();
```
This won't work, because you are caching the value, instead it needs to be read every time since PowerRename can enabled/disabled at any time in the settings.
Also, since it won't be acceptable to parse the whole json every time just for reading the `enable` state, until we find a better solution, we should keep using the registry for it.
ImageResizer has the same issue and we need to come up with the same solution for both.

vldmr11080 replied:
But `CSettings` object will exist no matter if `PowerRename` is disabled in the settings, and You can still read it from there. When we disable `PowerRename` we don't destroy if, but just put enabled flag and save settings. If we still need to be able to read this even when entire `PowerToys` applications is destroyed, than we need to change to registry. Otherway, this will work as long also `PowerToys` exist, no matter if `PowerRename` is enabled / disabled.

enricogior replied:
The problem is that `CSettings` is caching the enabled value and if you update the settings in the UI the new value is not used by this code.
You have to not cache the enabled value and read it every time.

vldmr11080 replied:
We update json whenever we enable/disable from UI, also we update this "cached" value, so it's always up to date:
```
    void save_settings()
    {
        CSettingsInstance().SetEnabled(m_enabled);
        CSettingsInstance().SavePowerRenameData();
        Trace::EnablePowerRename(m_enabled);
    }
```

enricogior replied:
It doesn't work because you don't read it again while the DLL is loaded in memory by Explorer. If you build the MSI, install it and test it, you'll see that it doesn't work.
Also, as I already said, since we should not parse the whole json every time we need to check the enabled value, we should not store it in json. ImageResizer has the same problem, we need to figure out a solution for both. In the meantime, please use the registry to persist the enabled state and make sure to read it every time it is used, as it was before this change.

enricogior commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,113 @@ void CRenameMRU::_FreeMRUList()
+{
+    json::JsonObject jsonData;
+
+    jsonData.SetNamedValue(c_enabled,                 json::value(settings.enabled));
```
As per previous comment, remove this line since we don't save `enabled` in the json.

vldmr11080 replied:
Removed.

enricogior commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,113 @@ void CRenameMRU::_FreeMRUList()
+
+void CSettings::MigrateSettingsFromRegistry()
+{
+    settings.enabled                 = GetRegBoolean(c_enabled, true);
```
Also remove this line.

vldmr11080 replied:
Removed.

enricogior commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,113 @@ void CRenameMRU::_FreeMRUList()
+    if (json)
+    {
+        const json::JsonObject& jsonSettings = json.value();
+        if (json::has(jsonSettings, c_enabled, json::JsonValueType::Boolean))
+        {
+            settings.enabled = jsonSettings.GetNamedBoolean(c_enabled);
+        }
```
Remove these lines.

vldmr11080 replied:
Removed.

enricogior commented on src/modules/powerrename/lib/Settings.cpp:
```diff
@@ -470,12 +359,113 @@ void CRenameMRU::_FreeMRUList()
+        }
+        if (json::has(jsonSettings, c_showIconOnMenu, json::JsonValueType::Boolean))
+        {
+            settings.showIconOnMenu = jsonSettings.GetNamedBoolean(c_showIconOnMenu);
```
`GetNamedBoolean` and the other json get methods throw exceptions, we need to catch them.

vldmr11080 replied:
You are right, before I was using the ones with default parameter which don't throw exception, so I forgot to add it now. Will change.

enricogior replied:
But `GetNamedBoolean` is still a method can could potentially throw an exception, so we have to catch it even if we reduce the chance to happen.

vldmr11080 replied:
Added try-catch block.

enricogior commented on src/modules/powerrename/lib/Settings.h:
```diff
@@ -1,45 +1,131 @@
+
+    inline bool GetEnabled() const
+    {
+        return settings.enabled;
```
Read from the registry.

vldmr11080 replied:
Done.

enricogior commented on src/modules/powerrename/lib/Settings.h:
```diff
@@ -1,45 +1,131 @@
-    static bool GetRegStringValue(_In_ PCWSTR valueName, __out_ecount(cchBuf) PWSTR value, DWORD cchBuf);
+    struct Settings
+    {
+        bool enabled{ true };
```
Remove it.

vldmr11080 replied:
Removed.

vldmr11080 committed:
Don't migrate enabled flag. Always read / write from registry.

enricogior commented:
@yuyoyuppe PTAL, thanks.

enricogior commented:
The ReplaceMRU and SearchMRU lists are still saved to the registry.

![image](https://user-images.githubusercontent.com/3206696/78793763-24b26780-79b3-11ea-8abc-2319f5b71cf8.png)

vldmr11080 commented:
> The ReplaceMRU and SearchMRU lists are still saved to the registry.
> 
> ![image](https://user-images.githubusercontent.com/3206696/78793763-24b26780-79b3-11ea-8abc-2319f5b71cf8.png)

Migration of most recently used search and replace strings will be implemented using json storage as well (instead of registry) in follow up pull request.
